src/nano_zlib.cc

```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
+          (last_bytes_ & 0xff) == ((unemitted_bytes_ >> 24) & 0xff)) {
+        // We recognized that the three first unemitted bytes are repetitions
+        // of the last emitted one.  It's not a huge win to code them as
+        // a run of 1+3, but it saves a few bits.
+        mode_ = REP1;
+        repetitions_ = 3;
+        last_bytes_ <<= 24;
+        last_bytes_ |= unemitted_bytes_ >> 8;
+        last_bytes_valid_ = 4;
+        unemitted_bytes_ &= 0xff;
+        unemitted_bytes_valid_ = 1;
+        output_repetitions();  // Will flush completely since repetitions_ == 3.
+      } else if (last_bytes_valid_ >= 4 &&
+          last_bytes_ == unemitted_bytes_) {
+        // We recognized that the four unemitted bytes are the same as
+        // the last four emitted ones.  Start a run with offset 4.
+        mode_ = REP4;
+        repetitions_ = 4;
+        unemitted_bytes_ = 0;
+        unemitted_bytes_valid_ = 0;
+      } else if (last_bytes_valid_ >= 3 &&
+          (last_bytes_ & 0xffffff) == (unemitted_bytes_ >> 8)) {
+        // We recognized that the first four of the unemitted bytes are
+        // the same as the last three emitted ones.
+        mode_ = REP3;
+        if ((unemitted_bytes_ & 0xff) == ((last_bytes_ >> 16) & 0xff)) {
+          // The last unemitted byte also matches the pattern of three
+          // being repeated.
+          last_bytes_ = unemitted_bytes_;
+          last_bytes_valid_ = 4;
+          repetitions_ = 4;
+          unemitted_bytes_ = 0;
+          unemitted_bytes_valid_ = 0;
+        } else {
+          // The last unemitted byte didn't match, so it was a rather
+          // short run.
+          repetitions_ = 3;
+          last_bytes_ <<= 24;
+          last_bytes_ |= unemitted_bytes_ >> 8;
+          unemitted_bytes_ &= 0xff;
+          unemitted_bytes_valid_ = 1;
+          output_repetitions();  // Will flush completely since repetitions_ == 3.
+        }
+      }
+    }
+    if (mode_ == LITERAL) {
+      if (unemitted_bytes_valid_ == 4) {
+        // We have no current run, so just emit the oldest unemitted byte to
+        // make space for the new bytes.
+        uint8 to_emit = unemitted_bytes_ >> 24;
+        literal(to_emit);
+        unemitted_bytes_ <<= 8;
+        unemitted_bytes_ |= b;
+        last_bytes_ <<= 8;
+        last_bytes_ |= to_emit;
+        if (last_bytes_valid_ < 4) last_bytes_valid_++;
+      } else {
+        // We have no current run, but there's a spare slot in the unemitted
+        // bytes so we put our new byte there.
+        unemitted_bytes_ <<= 8;
+        unemitted_bytes_valid_++;
+        unemitted_bytes_ |= b;
+      }
     } else {
-      while (repetitions_ != 0) output_repetitions();
-      last_byte_ = b;
-      literal(b);
+      // We have some sort of run going.  As long as it is running we don't
+      // have unemitted bytes.
+      ASSERT(unemitted_bytes_valid_ == 0);
+      int shift = static_cast<int>(mode_ - 1) << 3;
```

Comment:
Add explicit values to the Mode constants and add a comment explaining that their value is crucial.
Alternatively make a map from mode to shift. (Almost prefer that).